README.md
[![Development strategy](https://img.shields.io/static/v1?label=DEVELOPMENT%20STRATEGY&message=GITHUB%20FLOW&color=blue)](https://docs.github.com/en/get-started/quickstart/github-flow)

# salesforce-integration #

Salesforce-Integration is a micro service meant to be a middleware that will allow communication between users of WhatsApp or Facebook bots and Coppel's human advisor through a live chat on the Salesforce platform.

The requirements of this solution are documented in the following [RFC-155.](docs/RFC-155.md).

We can have more detailed information on the flow of this solution in the following documentation:

- [Salesforce Integration Middleware](/docs/Salesforce-Integration-Middleware.md)

## What is this repository for? ##

This service will serve as a middleware that will connect a user of a Yalo bot when requesting assistance with a human
agent through a live chat on the Salesforce platform.

This solution aims to be used for different implementations with any client in a simple way, in which it is necessary to
establish a live chat, through a Yalo bot and the Salesforce CRM platform.

The flow implemented for creating a chat between an end user and a human agent in Salesforce is as follows:

1. The user is interacting with a Yalo bot.
2. If the user requests human assistance by typing the keyword ***"Ayuda"***, Yalo Core, puts the artificial
intelligence in a waiting state and Yalo Component creates a Chat session with Salesforce by making a request to the
endpoint
[`v1/chats/connect`](/docs/Salesforce-Integrations-Endpoints.md) of
this solution.
3. The endpoint must consume the following minimum information to create a chat:

| Field       | Description                                                                                                                                                                                                 |
|:------------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| userID      | User identifier phone in whatsapp and facebook Id in messenger.                                                                                                                                             |
| botSlug     | Bot name.                                                                                                                                                                                                   |
| botId       | Phone for whatsapp bot or pageId in messenger bot.                                                                                                                                                          |
| name        | User name.                                                                                                                                                                                                  |
| provider    | Chat origin, the allowed values are **whatsapp** or **facebook**.                                                                                                                                           |
| email       | User's email to search or register the contact or person account in Salesforce.                                                                                                                             |
| phoneNumber | User's phone number to search or register the contact or personal account in Salesforce allowed in whatsapp bot, only if it is enabled.                                                                     |
| extraData   | It mainly sends the custom fields that the customer has in Salesforce to add them to the cases and they are sent if we have the information. If necessary you can send metadata for custom implementations. |

4. First, it is validated that the user does not have a live chat in existence at the time of making the request.

5. We search through the email or phone received that there is a **contact** or **personal account** in the customer's
Salesforce account, if it does not exist, the contact or personal account is created.

6. We validate that the user is not blocked in Salesforce, if so, the corresponding chat is not created and through the
botrunner or studiong client the user is changed to a *blocked user state*, specified in the environment variable ***BLOCKED_USER_STATE***

7. Once the contact has been validated in Salesforce, we create a query case with the data provided to add it and
request a chat.
[ChatRequest in LiveAgent API](https://developer.salesforce.com/docs/atlas.en-us.live_agent_rest.meta/live_agent_rest/live_agent_rest_ChasitorInit.htm)

8. After successfully requesting the chat session with Salesforce, we save the information necessary to maintain the
session in an **Interconnection** object in Redis.

9. Finally, we initialize in the ***Interconnection*** a goroutine that starts the
[***Long Polling***](https://developer.salesforce.com/docs/atlas.en-us.live_agent_rest.meta/live_agent_rest/live_agent_rest_http_long_polling_loop.htm)
process with Salesforce, with which we consume a service to identify the different events that happen in the chat every
5 seconds.

10. According to the events received, **salesforce-integration** performs the following actions:

* ***204 HTTP Status :*** It means that there are no changes in the chat.
* ***503 HTTP Status :*** An attempt will be made to reconnect the session because the connection to Salesforce was
lost. More details [ReconnectSession](https://developer.salesforce.com/docs/atlas.en-us.live_agent_rest.meta/live_agent_rest/live_agent_rest_ReconnectSession.htm)
* ***Other HTTP Status:*** The **Interconnection** is ended and through the **Botrunner or Studio NG client** the user
is changed to a time-ended state, specified in the **TIMEOUT_STATE** environment variable
* ***ChatRequestFail:*** The chat was not created correctly in the Salesforce console, due to no agents available in the queue the chat was sent to or some unknown error.
* ***ChatRequestSuccess :*** The chat was created successfully and the **Interconnection** goes to `OnHold` status waiting for a human agent to accept the chat. The service sends the user a `Esperando agente` message via **Integrations API**
* ***ChatEstablished :*** An agent accepts the chat and the **Interconnection** goes to `Active` status, at this time both can send messages to each other, but before that, the service sends the context of the query as the first message, that is, the previous conversation between the bot's AI and the user before requesting a live chat.
* ***ChatMessage :*** The agent sent a message to the user, then through **Integrations API** we forward the message to the user to their *WhatsApp* or *Messenger*.
* ***ChatEnded :*** The agent ended the chat and the **Interconnection** goes to `Closed` status, closing the session in the integration and through the **Botrunner or Studiong client** the user is changed to a successfully *finished chat state*, specified in the environment variable **SUCCESS_STATE**.

***How do I send a user's response to an agent in the Salesforce console?***

When the instance is raised, we need to register the webhooks through the endpoint `/v1/integrations/webhook/register/{{provider}}`, then the webhooks are registered in the Integrations API that will receive the messages between the users and the bot.

If we receive a text or image message from a user, we validate the following:

* If the user does not have an Active or OnHold chat, this message is saved as context in Redis.

* If there is an Active chat then through our [LiveAgent API client](https://developer.salesforce.com/docs/atlas.en-us.live_agent_rest.meta/live_agent_rest/live_agent_rest_ChatMessage_request.htm) we send that message to the agent's console. If the message is of type Image, a message is sent to the agent saying `"El usuario adjunto una imagen al caso"`, and the image through our Salesforce API client is attached to the case.

Any other type of message other than TEXT or IMAGE in the webhooks is ignored.

The service has two main folders:

## App folder ##

This folder contains the main functionalities of the service together with the packages for the API server.

The **api/hadlers** package contains the endpoints available for this integration, mainly the endpoint
***/chats/connect*** to create the chats with Salesforce, as well as the endpoints that we need to register as webhooks
in **integrations-api** so that **integrations-channels** can send us the messages between users and bots of *WhatsApp*
and *Faceebook*.

Each of them is detailed in the [README](/app/api/handlers/README.md) file in ***app/api/handlers*** or in [Saleforce Integrations Endpoints](/docs/Salesforce-Integrations-Endpoints.md).

In **/manager** folder we will find all the logic implemented to be able to manage the chats between the end users and the agents in Salesforce. For this we explain some components that we need to be able to perform these functions:

* ***interconnectionMap:*** This map is a local cache that will store the **Interconnections** on ONHOLD and ACTIVE. Once the chats are finished they are removed from this map and the `long-polling` service with Salesforce is closed.

* [***Interconnection:***](/docs/Interconnection.md) Structure that represents a live chat session between a Yalo bot user and a human agent in the Salesforce console. This object is stored in Redis and contains all the necessary information to be able to start the `Long-Polling` service with Salesforce. In the following table we see the necessary information that is needed and stored in Redis:

| Field         | Description                                                                                                            |
|:--------------|:-----------------------------------------------------------------------------------------------------------------------|
| userID        | User identifier phone in whatsapp and facebook Id in messenger.                                                        |
| sessionId     | Chat session identifier in Salesforce                                                                                  |
| sessionKey    | Security key for the chat session required for requests.                                                               |
| affinityToken | The affinity token for the session that’s passed in the header for all future requests.                                |
| status        | The three states are: OnHold, Active and Closed.                                                                       |
| timestamp     | Timestamp of the creation of the inteconnection.                                                                       |
| botSlug       | Bot name.                                                                                                              |
| botId         | Phone for whatsapp bot or pageId in messenger bot.                                                                     |
| name          | User name.                                                                                                             |
| provider      | Chat origin, the allowed values are **whatsapp** or **facebook**.                                                      |
| email         | Email received to create the chat.                                                                                     |
| phoneNumber   | Phone received to create the chat.                                                                                     |
| caseId        | The ID of the case created in Salesforce for this chat session.                                                        |
| extraData     | The custom fields that the customer has in Salesforce to add them to the cases or metadata for custom implementations. |

* ***contextcache:*** This redis client will store the context of each user of all the messages sent between him and the bot, received through the webhook.

* ***context:*** The context is the messages sent between the user and the bot before a live chat is requested. This struct represents the information that we store in redis whose time-to-live is only 24 hours.

| Field     | Description                                                     |
|:----------|:----------------------------------------------------------------|
| userID    | User identifier phone in whatsapp and facebook Id in messenger. |
| timestamp | Timestamp of sending the message.                               |
| url       | Image url if it is an image type message.                       |
| mimeType  | MIME Type of the image.                                         |
| text      | Message text.                                                   |
| from      | Who sent the message the user or the bot.                       |

* ***SalesforceService:*** This component contains the clients to connect to the Salesforce APIs and the functions used with Salesforce:
  1. Authorization API
  2. [Live Agent Rest API](https://developer.salesforce.com/docs/atlas.en-us.live_agent_rest.meta/live_agent_rest/live_agent_rest_API_requests.htm)
  3. [Data Salesforce API](https://developer.salesforce.com/docs/atlas.en-us.api.meta/api/sforce_api_quickstart_intro.htm)

* ***IntegrationsClient:*** Client to use message sending and webhook logging endpoints with [Integrations API](https://developer.yalo.com/yalo/reference/integrations-api-overview).

* ***BotrunnnerClient:*** SentTo client to be able to change a user's status in the bot flow.

#### Base folder

This folder has all the support packages that will be used by the app folder.

Here is the list of packages in the base folder with its explanation:

| Package   | Description                                                                                                                         | Implementations or sub-packages                                                                                           |
|-----------|-------------------------------------------------------------------------------------------------------------------------------------|---------------------------------------------------------------------------------------------------------------------------|
| cache     | Contains logic to retrieve and store mensages of context and interconnections on cache.                                             | Implementations: Redis (ContextCache, InterconnectionCache) and [Ristretto Cache](https://github.com/dgraph-io/ristretto) |
| clients   | Contains the clients to send requests to Botrunner sent-to, Studiong sent-to, Integrations API, AgentLiveChat API and Saleforce API |                                                                                                                           |
| constants | It will contain the common constants and errors that we use in the service                                                          |                                                                                                                           |
| helpers   | It contains utilities that help us in the creation of endpoints or file encoding, and other functionalities.                        |                                                                                                                           |
| models    | Contains base structures that we will use for this service.                                                                         |                                                                                                                           |

## How do I get set up? ##

* For this service you don't need to have a REDIS DB up and running. It's not necessary to have a an specific version or replication strategy.
* It is necessary to create a Json Web Token through the `/authenticate`  endpoint, sending a username and password in the body of the request.
* This integration can change the state of a user in the bot through botrunner or studiong, you just have to add the ENV API vars you want to use.
* This integration can send messages to the bot through integrations API.
* Here are the required ENV vars needed to run this service:

| Name                                                  | Description                                                                                                                                                                                                                                                                                     | Required                                        | Defaults                                          |
|-------------------------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-------------------------------------------------|---------------------------------------------------|
| SALESFORCE-INTEGRATION_APP_NAME                       | The connection name displayed on the application.                                                                                                                                                                                                                                               | true                                            | salesforce-integration                            |
| SALESFORCE-INTEGRATION_HOST                           | The host for the service API.                                                                                                                                                                                                                                                                   | true                                            | localhost                                         |
| SALESFORCE-INTEGRATION_PORT                           | The port for the service API.                                                                                                                                                                                                                                                                   | true otherwise 8080                             | 8080                                              |
| SALESFORCE-INTEGRATION_SENTRY_DSN                     | The project DSN for Sentry events.                                                                                                                                                                                                                                                              | false                                           |                                                   |
| SALESFORCE-INTEGRATION_ENVIRONMENT                    | The environment used for Sentry events or also for some functions that are required in dev or prod environment.                                                                                                                                                                                 | false                                           | dev                                               |
| SALESFORCE-INTEGRATION_MAIN_CONTEXT_TIME_OUT          | The time out for the main context in seconds.                                                                                                                                                                                                                                                   | false                                           | 10                                                |
| SALESFORCE-INTEGRATION_REDIS_ADDRESS                  | The address of the Redis Instance (not sentinel).                                                                                                                                                                                                                                               | Only if Redis Instance is not a sentinel Redis. |                                                   |
| SALESFORCE-INTEGRATION_REDIS_PASSWORD                 | The password of the Redis Instance.                                                                                                                                                                                                                                                             | Only if Redis Instance is not a sentinel Redis. |                                                   |
| SALESFORCE-INTEGRATION_REDIS_MASTER                   | The name of the master on a Redis Sentinel Instance (sentinel).                                                                                                                                                                                                                                 | Only if Redis Instance is a sentinel Redis.     |                                                   |

CHANGELOG.md

<a name="v2.1.0"></a>
## [v2.1.0](https://github.com/yalochat/salesforce-integration/compare/v2.1.0..v2.0.3) (2023-01-16)

### Chore

* **Documentation:** Removing sensible information from the files. Close #CS-2216
* **Documentation:** Adding two missing diagrams to the README.md file. Close #CS-2216
* **Documentation:** prepare documentation into the repo to have it available with no other deps. Close #CS-2216

### Feat

* **salesforce_service:** add the possibility to send the AccountRecordTypeId on extraData
* **version:** upgrade version for changes in documentation. Closes #CS-2216

### Fix

* **changelog:** fix changelog


<a name="v2.0.3"></a>
## [v2.0.3](https://github.com/yalochat/salesforce-integration/compare/v2.0.3..v2.0.2) (2023-01-10)

### Fix

* change to log warn when could not find a contact
* stop sending span error to DD if didn't find the contact
* **integrations-request:** stop loggin info for every ignored message
* **interconnection_test:** fix flaky interconnection test
* **long-polling-loop:** increase long polling timeout to 60s
* **manager:** remove interconnection from logrus
* **manager:** update regex that defines the attach name
* **search-contact:** stop loggin error in case had not found the contact
* **sfc_data:** update DD tag to be a slug
* **webhook_handler:** fix DD span


<a name="v2.0.2"></a>
## [v2.0.2](https://github.com/yalochat/salesforce-integration/compare/v2.0.2..v2.0.1) (2023-01-09)

### Feat

* **searchContact:** update service to search contacts by custom fields


<a name="v2.0.1"></a>
## [v2.0.1](https://github.com/yalochat/salesforce-integration/compare/v2.0.1..v2.0.0) (2023-01-06)

### Hotfix

* **long-polling:** add support for initial sequence number equals to zero on GetMessages operation. Closes #CS-2220


<a name="v2.0.0"></a>
## [v2.0.0](https://github.com/yalochat/salesforce-integration/compare/v2.0.0..v1.1.8) (2023-01-05)

### Feature

* **get-messages:** adds the ack query parameters to get message requests. Closes #CS-2170


<a name="v1.1.8"></a>
## [v1.1.8](https://github.com/yalochat/salesforce-integration/compare/v1.1.8..v1.1.7) (2022-08-18)


<a name="v1.1.7"></a>
## [v1.1.7](https://github.com/yalochat/salesforce-integration/compare/v1.1.7..v1.1.6) (2022-06-01)

### Hotfix

* **webhook integrations:** Add the possibility to filter the messages types


<a name="v1.1.6"></a>
## [v1.1.6](https://github.com/yalochat/salesforce-integration/compare/v1.1.6..v1.1.5) (2022-05-02)

### Feat

* **Integration - Interconnection:** Decrease wait time in long polling. Decrease wait time in long polling  between each request to salesforce. Closes #UI-53
* **Integration - Interconnection:** Add pprof endpoints. Add profile endpoints to check api performance. Closes #UI-52


<a name="v1.1.5"></a>
## [v1.1.5](https://github.com/yalochat/salesforce-integration/compare/v1.1.5..v1.1.4) (2022-03-31)

### Feat

* **Integration-interconnection:** Implement redis enterprise on salesforce and s1 | Staging. Add redis enterprise config connection. Closes #UI-27.

### Fix

* **Integration-interconnection:** Hotfix redis salesforce. Omit error redis nil to create chat. Closes #UI-37


<a name="v1.1.4"></a>
## [v1.1.4](https://github.com/yalochat/salesforce-integration/compare/v1.1.4..v1.1.3) (2022-03-25)

### Feat

* **General improvemnts:** Show text when an image contains it. Closes #UI-36
* **Integration-interconnection:** End chat correctly. Change validation EndChat endpoint to validate interconnection in redis. Closes #UI-33
* **Integration-interconnection:** Improve handling redis errors. Add error redis handler in validate user function. Closes #UI-32


<a name="v1.1.3"></a>
## [v1.1.3](https://github.com/yalochat/salesforce-integration/compare/v1.1.3..v1.1.2) (2022-03-02)

### Feat

* **Improve Salesforce implementation:** Upload files on cases. Upload files and add them to the case when there is an active chat and an attachment is sent on whatsapp and facebook. Closes #UI-30


<a name="v1.1.2"></a>
## [v1.1.2](https://github.com/yalochat/salesforce-integration/compare/v1.1.2..v1.1.1) (2022-03-02)


<a name="v1.1.1"></a>
## [v1.1.1](https://github.com/yalochat/salesforce-integration/compare/v1.1.1..v1.1.0) (2022-02-15)

### Feat

* **Improve Salesforce implementation:** Add reconnect on long-polling. Do a reconnection when a 503 http error occurs on long polling. Closes #UI-25
* **Integration-interconnection:** Redis singleton connection Salesforce. Change CreateManager to Redis singleton connection. Closes #UI-23.
* **Integration-interconnection:** Implement kafka. Add Kafka to the project to avoid lost messages. Closes #UI-8.

### Fix

* **Improve Salesforce implementation:** Adding producer to interconnections in restore. Change the order of the Kakfa producer declaration, to be done before restoring interconnects. Closes #UI-26


<a name="v1.1.0"></a>
## [v1.1.0](https://github.com/yalochat/salesforce-integration/compare/v1.1.0..v0.1.9) (2022-02-03)

### Feat

* **Improve Salesforce implementation:** Fix error message on Salesforce clients. Correctly sending the response sent by the salesforce API on the errors returned, including datadog traces. Closes #UI-19
* **Improve Salesforce implementation:** Add userID and client on datadog traces Salesforce. Adding labels with values ​​such as sessionID, userID and client to more easily find errors in the traces. Closes #UI-18
* **Improve Salesforce implementation:** Add SIGTERM. The service must be able to close connections correctly when a shutdown occurs or the app ends. Closes #UI-6
* **Improve Salesforce implementation:** Avoid status messages on webhooks. We avoid status messages that integrations channels send when the bot is configured to send them. Closes #UI-10
* **Improve Salesforce implementation:** Handle errors on /chats/connect. Send a sentTo to the user when an error occurs when creating a chat with Salesforce. Closes #UI-2

app/config/envs/envs.go
package envs

import (
	"encoding/json"
	"fmt"
	"strings"
	"time"

	"yalochat.com/salesforce-integration/base/models"
)

// Envs represents the list of well known env vars used by the app
type Envs struct {
	AppName                        string                 `default:"salesforce-integration" split_words:"true"`
	Client                         string                 `default:"salesforce" split_words:"true"`
	Host                           string                 `required:"true" split_words:"true" default:"localhost"`
	Port                           string                 `required:"true" split_words:"true" default:"8080"`
	SentryDSN                      string                 `default:"" split_words:"true"`
	Environment                    string                 `default:"dev" split_words:"true"`
	RedisAddress                   string                 `split_words:"true"`
	RedisMaster                    string                 `split_words:"true"`
	RedisPassword                  string                 `split_words:"true"`
	RedisSentinelAddress           string                 `split_words:"true"`
	BlockedUserState               map[string]string      `required:"true" split_words:"true" default:"whatsapp:from-sf-blocked,facebook:from-sf-blocked"`
	TimeoutState                   map[string]string      `required:"true" split_words:"true" default:"whatsapp:from-sf-timeout,facebook:from-sf-timeout"`
	SuccessState                   map[string]string      `required:"true" split_words:"true" default:"whatsapp:from-sf-success,facebook:from-sf-success"`
	YaloUsername                   string                 `required:"true" split_words:"true" default:"yaloUser"`
	YaloPassword                   string                 `required:"true" split_words:"true"`
	SalesforceUsername             string                 `required:"true" split_words:"true" default:"salesforceUser"`
	SalesforcePassword             string                 `required:"true" split_words:"true"`
	SecretKey                      string                 `required:"true" split_words:"true"`
	BotrunnerUrl                   string                 `split_words:"true"`
	BotrunnerToken                 string                 `split_words:"true" default:""`
	BotrunnerTimeout               int                    `split_words:"true" default:"4"`
	StudioNGUrl                    string                 `split_words:"true"`
	StudioNGToken                  string                 `split_words:"true"`
	StudioNGTimeout                int                    `split_words:"true" default:"4"`
	SfcClientID                    string                 `split_words:"true"`
	SfcClientSecret                string                 `split_words:"true"`
	SfcUsername                    string                 `split_words:"true"`
	SfcPassword                    string                 `split_words:"true"`
	SfcSecurityToken               string                 `split_words:"true"`
	SfcBaseUrl                     string                 `split_words:"true"`
	SfcChatUrl                     string                 `split_words:"true"`
	SfcLoginUrl                    string                 `split_words:"true"`
	SfcApiVersion                  string                 `split_words:"true" default:"52"`
	SfcOrganizationId              string                 `split_words:"true"`
	SfcDeploymentId                string                 `split_words:"true"`
	SfcRecordTypeId                string                 `split_words:"true"`
	SfcAccountRecordTypeId         string                 `split_words:"true"`
	SfcDefaultBirthDateAccount     string                 `split_words:"true" default:"1921-01-01T00:00:00"`
	SfcCustomFieldsCase            map[string]string      `split_words:"true"`
	SfcCustomFieldsContact         map[string]string      `split_words:"true"`
	SfcCustomFieldsAccount         map[string]string      `split_words:"true"`
	SfcSourceFlowBot               SfcSourceFlowBot       `required:"true" split_words:"true"`
	SfcSourceFlowField             string                 `required:"true" split_words:"true" default:"source_flow_bot"`
	SfcBlockedChatField            bool                   `split_words:"true" default:"false"`
	SfcCodePhoneRemove             []string               `split_words:"true" default:"521,52"`
	IgnoreMessageTypes             string                 `split_words:"true" default:"status,notification-status"`
	IntegrationsWAChannel          string                 `split_words:"true" default:"outgoing_webhook"`
	IntegrationsFBChannel          string                 `split_words:"true" default:"passthrough"`
	IntegrationsWABotID            string                 `split_words:"true"`
	IntegrationsFBBotID            string                 `split_words:"true"`
	IntegrationsWABotJWT           string                 `split_words:"true"`
	IntegrationsFBBotJWT           string                 `split_words:"true"`
	IntegrationsBaseUrl            string                 `split_words:"true"`
	IntegrationsSignature          string                 `split_words:"true"`
	WebhookBaseUrl                 string                 `split_words:"true"`
	IntegrationsWABotPhone         string                 `split_words:"true"`
	IntegrationsFBBotPhone         string                 `split_words:"true"`
	KeywordsRestart                []string               `split_words:"true" default:"coppelbot,regresar,reiniciar,restart"`
	SpecSchedule                   string                 `split_words:"true" default:"@every 59m"`
	MaxRetries                     int                    `split_words:"true" default:"2"`
	CleanContextSchedule           string                 `split_words:"true" default:"0 9 * * *"`
	IntegrationChanRateLimit       float64                `split_words:"true" default:"20"`
	SaleforceChanRateLimit         float64                `split_words:"true" default:"20"`
	Messages                       models.MessageTemplate `split_words:"true" required:"true" default:"{\"waitAgent\":\"Esperando un agente\",\"welcomeTemplate\":\"Hola soy %s y necesito ayuda\",\"context\":\"Contexto\",\"DescriptionCase\":\"Caso levantado por el Bot\",\"uploadImageError\":\"Imagen no enviada\",\"uploadImageSuccess\":\"**El usuario adjunto una imagen al caso**\",\"uploadFileError\":\"Archivo no enviado\",\"uploadFileSuccess\":\"**El usuario adjunto un archivo al caso**\",\"queuePosition\":\"Posici\u00F3n en la cola\",\"waitTime\":\"Tiempo de espera\",\"firstNameContact\":\"Contacto Bot - \",\"clientLabel\":\"Cliente\",\"botLabel\":\"Bot\",\"replyToTemplate\":\"Respuesta de => [%s] \\n \\n Mensaje enviado => %s\",\"agentLabel\":\"Agente\"}"`
	Timezone                       string                 `required:"true" default:"America/Mexico_City"`
	SendImageNameInMessage         bool                   `split_words:"true" default:"false"`
	KafkaHost                      string                 `required:"true" split_words:"true"`
	KafkaPort                      string                 `required:"true" split_words:"true"`
	KafkaUser                      string                 `required:"true" split_words:"true"`
	KafkaPassword                  string                 `required:"true" split_words:"true"`
	KafkaTopic                     string                 `required:"true" split_words:"true"`
	UseProfile                     bool                   `split_words:"true" default:"false"`
	SleepLongPollling              time.Duration          `split_words:"true" default:"3s"`
	SfcCustomFieldsToSearchContact map[string]string      `split_words:"true"`
}

type Provider struct {
	ButtonID string `json:"button_id"`
	OwnerID  string `json:"owner_id"`
}

type SourceFlowBot struct {
	Subject   string              `json:"subject"`
	Providers map[string]Provider `json:"providers"`
}

type SfcSourceFlowBot map[string]SourceFlowBot

//Decode Decoder this function deserializes the struct by the envconfig Decoder interface implementation
func (sd *SfcSourceFlowBot) Decode(value string) error {
	providerMap := map[string]SourceFlowBot{}

	pairs := strings.Split(value, ";")
	for _, pair := range pairs {
		sourceFlowBotData := SourceFlowBot{}
		kvpair := strings.Split(pair, "=")
		if len(kvpair) != 2 {
			return fmt.Errorf("invalid map item: %q", pair)
		}

		err := json.Unmarshal([]byte(kvpair[1]), &sourceFlowBotData)
		if err != nil {
			return fmt.Errorf("invalid map json: %w", err)
		}

		providerMap[kvpair[0]] = sourceFlowBotData

	}
	*sd = SfcSourceFlowBot(providerMap)

	return nil
}
